src/commands/plugins.rs

```diff
@@ -242,11 +243,99 @@ impl Upgrade {
 
         if self.all {
             self.upgrade_all(manifests_dir).await
+        } else if self.name.is_none()
+            && self.local_manifest_src.is_none()
+            && self.remote_manifest_src.is_none()
+        {
+            // Default behavior (multiselect)
+            self.upgrade_multiselect().await
         } else {
             self.upgrade_one().await
         }
     }
 
+    // Multiselect plugin upgrade experience
+    async fn upgrade_multiselect(self) -> Result<()> {
+        let catalogue_plugins = list_catalogue_plugins().await?;
+        let installed_plugins = list_installed_plugins()?;
+
+        let installed_in_catalogue: Vec<_> = installed_plugins
+            .into_iter()
+            .filter(|installed| {
+                catalogue_plugins.iter().any(|catalogue| {
+                    installed.manifest == catalogue.manifest
+                        && matches!(catalogue.compatibility, PluginCompatibility::Compatible)
+                })
+            })
+            .collect();
+
+        let mut eligible_plugins = Vec::new();
+
+        // Getting only eligible plugins to upgrade
+        for installed_plugin in installed_in_catalogue {
+            let manager = PluginManager::try_default()?;
+            let manifest_location = ManifestLocation::PluginsRepository(PluginLookup::new(
+                &installed_plugin.name,
+                None,
+            ));
+
+            // Attempt to get the manifest to check eligibility to upgrade
+            if let Ok(manifest) = manager
+                .get_manifest(&manifest_location, false, SPIN_VERSION)
+                .await
+            {
+                if installed_plugin.version != manifest.version() {
```

Comment:
In theory we should check that the catalogue manifest version is _greater than_ the installed version but I don't know if that's ever an issue in practice.

Reply:
Yeah, you're right, even if this is an issue in practice, in theory, we should check if the version is greater than the installed one.
What we could do is parse the strings to numbers and compare them. What do you think?

Reply:
If we were to do this, we would not parse to numbers but to the `semver::Version` type (see https://docs.rs/semver/latest/semver/).  The `Manifest` type has a `try_version()` method which can give you this  And `semver::Version` allows you to use comparison operators such as `<` and `>` (see e.g. https://github.com/fermyon/spin/blob/1d662e3774f85106de4b212ae48d0d4760f47ee2/crates/plugins/src/badger/mod.rs#L285).

That said, it's up to you whether you want to wrestle with this.  Unfortunately, it is not quite as easy as just calling `try_version()` and doing a comparison, because `try_version` can return an error (if the version is malformed).  Which means we have to think about error handling.  Rust makes it very easy to simply bomb out (via the `?` operator), but that could mean one bad version string broke the whole experience.  One option would be to write a helper function that handled the error cases by falling back to string comparison, along the lines of

```
fn is_potential_upgrade(current: &PluginManifest, candidate: &PluginManifest) -> bool {
  match (current.try_version(), candidate.try_version()) {
    (Ok(cur_ver), Ok(cand_ver)) => cand_ver > cur_ver,
    _ => current.version() != candidate.version(),
  }
}
```

and then this line would be `if is_potential_upgrade(&installed_plugin, &manifest)`.  _(Disclaimer: written in the GitHub comment box, not tested, unlikely to compile as is!)_

It's up to you whether you want to invest this amount of effort - it's not a blocker as far as I'm concerned.

Reply:
Why `current.version() != candidate.version()` if something goes wrong? I know this is the condition we have at the moment, but shouldn't we return false if something goes wrong with `try_version()`'s?

Reply:
Suppose we returned `false` if one of the versions failed to parse.  Now suppose that somehow (again, should never happen) the user's _installed_ version fails to parse.  Now they don't have the option to upgrade.

This could be avoided by having it return `false` if only the _target_ version fails to parse.  But since we're displaying the from and to versions, I'd recommend erring on the side of leaving the choice with the user.

Again, it should be moot because there should _never_ be a malformed version in the catalogue.  But that was the motivation for the more forgiving fallback.